lib/semantics/mod-file.cc

```diff
@@ -590,84 +608,116 @@ std::ostream &PutLower(std::ostream &os, const std::string &str) {
   return os;
 }
 
-// Write the module file at path, prepending header. Return false on error.
-static bool WriteFile(const std::string &path, std::string &&contents) {
-  std::fstream stream;
-  auto header{GetHeader(contents)};
-  auto size{GetFileSize(path)};
-  if (size == header.size() + 1 + contents.size()) {
-    // file exists and has the right size, check the contents
-    stream.open(path, std::ios::in | std::ios::out);
-    if (FileContentsMatch(stream, header, contents)) {
-      return true;
-    }
-    stream.seekp(0);
-  } else {
-    stream.open(path, std::ios::out);
+struct Temp {
+  Temp() = delete;
+  ~Temp() {
+    close(fd);
+    unlink(path.c_str());
+  }
+  int fd;
+  std::string path;
+};
+
+// Create a temp file in the same directory and with the same suffix as path.
+// Return an open file descriptor and its path.
+static Temp MkTemp(const std::string &path) {
+  auto length{path.length()};
+  auto dot{path.find_last_of("./")};
+  std::string suffix{dot < length && path[dot] == '.' ? path.substr(dot) : ""};
+  CHECK(length > suffix.length() &&
+      path.substr(length - suffix.length()) == suffix);
+  auto tempPath{path.substr(0, length - suffix.length()) + "XXXXXX" + suffix};
+  int fd{mkstemps(&tempPath[0], suffix.length())};
+  auto mask{umask(0777)};
+  umask(mask);
+  chmod(tempPath.c_str(), 0666 & ~mask);  // temp is created with mode 0600
+  return Temp{fd, tempPath};
+}
+
+// Write the module file at path, prepending header. If an error occurs,
+// return an error message, otherwise nullptr.
+static int WriteFile(const std::string &path, const std::string &contents) {
+  auto header{std::string{ModHeader::bom} + ModHeader::magic +
+      CheckSum(contents) + ModHeader::terminator};
+  if (FileContentsMatch(path, header, contents)) {
+    return 0;
+  }
+  Temp temp{MkTemp(path)};
+  if (temp.fd < 0) {
+    return errno;
+  }
+  if (write(temp.fd, header.c_str(), header.size()) == -1 ||
```

Comment:
It's better practice to check the return value against the byte count, so that short writes are caught as errors.

Reply:
OK, done. `read()` as well.